Sources/llbuild2/Core/KeyDependencyGraph.swift

```diff
@@ -0,0 +1,128 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+
+import Dispatch
+
+public enum LLBKeyDependencyGraphError: Error {
+    case cycleDetected([LLBKey])
+}
+
+/// Key -> Key dependency graph maintainer. Currently used for cycle detection during the request of keys.
+public class LLBKeyDependencyGraph {
+    // Use Ints containing the hashValue values for each key. They should be stable during the execution of this
+    // process, and we're not storing them for usage in future process invocations. If that changes, this should
+    // probably be refactored.
+    private var edges: [Int: Set<Int>]
+
+    // A map of the hashValue to the key. This allows us to preserve key information if paths are founds, so that
+    // they can be useful when debugging.
+    private var knownKeys: [Int: LLBKey]
+
+    private let queue = DispatchQueue.global(qos: .userInteractive)
+
+    public init() {
+        self.edges = [:]
+        self.knownKeys = [:]
+    }
+
+    /// Attempts to add a detected dependency edge to the graph, but throws if a cycle is detected.
+    public func addEdge(from origin: LLBKey, to destination: LLBKey) throws {
+        let originID = origin.digest.hashValue
+        let destinationID = destination.digest.hashValue
+
+        // Populate the knownKeys store with the new edge nodes, in case it doesn't exist.
+        queue.sync {
+            if self.knownKeys[originID] == nil {
+                self.knownKeys[originID] = origin
+            }
+            if self.knownKeys[destinationID] == nil {
+                self.knownKeys[destinationID] = destination
+            }
+        }
+
+        // Check if there's a path from the destination to the origin, as that would be a clear indication that adding
+        // the edge from the origin to the destination would introduce a cycle. This works because the graph starts
+        // by definition without cycles, so cycles can only be introduced if a new edge would create it.
+        if let path = anyPath(from: destinationID, to: originID) {
```

Comment:
This seems a bit unfortunate that we have to calculate this every time we add an edge.   Wish we could lazily check for a cycle if no progress is being made, like how llbuild prime does it.   Of course, thats a completely different engine design, so it's hard to translate here.

Reply:
yeah, agree with you... we already calculate the digest for every key during the build method in the engine, and I wish we had some sort of associated objects stuff like ObjC to store this once instead of calculating every time... Perhaps we can amortize this a bit by wrapping the key at the engine layer, and using that wrapper instead for the graph calculations...

re: the graph style for cycles, yeah, it's unfortunate we do it for every request... I'm not seeing a simpler method that maps into the futures model, though, but I'm ok with holding off on this  if you prefer to wait until we have a better idea on how to do this

Reply:
Probably worth having in for now. We can revisit it when it becomes a problem.